### PR TITLE
FC-1355 - Feature: trust http sigs for transactions

### PR DESCRIPTION
**Pull request by cap10morgan**
ledger counterpart to https://github.com/fluree/db/pull/147

JIRA ticket: FC-1355

Builds on #129 so review that one first :)

I implemented this by adding a new `:signed` flake type that holds the string we signed (to get the `:sig` flake) when that string is something other than the command in the `:tx` flake (which it previously always was before this PR).

### CODE REVIEW

File: src/fluree/db/ledger/consensus/update_state.clj [hunk 1]
```diff
@@ -72,31 +72,43 @@
                              current-val)))]
     (= proposed-v (get-in new-state ks))))
 
+(defn- extract-flake-object
+  "Returns flake object (.-o flake) from block-map of the given type whose
+  subject matches (:t tx-data). Returns nil if none is found."
+  [{:keys [flakes] :as _block-map} {:keys [t] :as _tx-data} type]
+  (some #(let [^Flake flake %]
```
Comment (zonotope) [inline src/fluree/db/ledger/consensus/update_state.clj:79]: you can use `flake/s`, `flake/p`, and `flake/o` below to avoid needing to carry around these type hints.

File: src/fluree/db/ledger/consensus/update_state.clj [hunk 2]
```diff
@@ -72,31 +72,43 @@
...
+             (filter #(and (= :new-db (:type (val %)))
+                           (= 200 (:status (val %)))))
+             (map (fn [[_ tx-data]]
+                    (let [efo           (partial extract-flake-object block-map
+                                                 tx-data)
+                          orig-cmd      (efo constants/$_tx:tx)
+                          orig-sig      (efo constants/$_tx:sig)
+                          orig-signed   (efo constants/$_tx:signed)
+                          orig-cmd-data (when orig-cmd (json/parse orig-cmd))
+                          {:keys [db fork forkBlock]} orig-cmd-data
```
Comment (zonotope) [inline src/fluree/db/ledger/consensus/update_state.clj:99]: you can combine lines 98 and 99

File: src/fluree/db/peer/http_api.clj [hunk 3]
```diff
@@ -206,19 +206,31 @@
...
-          _           (when-not (sequential? param)
-                        (throw (ex-info (str "A transaction submitted to the 'transact' endpoint must be a list/vector/array.")
-                                        {:status 400 :error :db/invalid-transaction})))
-          txid-only?  (some-> (get opts "txid-only") str/lower-case (= "true"))
-          auth-id     (:auth auth-map)
-          result      (<? (fdb/transact-async conn ledger param {:auth        auth-id
-                                                                 :private-key private-key
-                                                                 :txid-only   txid-only?
-                                                                 :timeout     timeout}))]
+    (let [conn          (:conn system)
```
Comment (zonotope) [inline src/fluree/db/peer/http_api.clj:209]: you can deref this in the argument vector

File: src/fluree/db/peer/http_api.clj [hunk 4]
```diff
@@ -206,19 +206,31 @@
...
+                              ;; original request
+                              (assoc :auth (:authority auth-map))))
+          _             (when-not (sequential? param)
+                          (throw (ex-info (str "A transaction submitted to the "
+                                               "'transact' endpoint must be a "
+                                               "list/vector/array.")
+                                          {:status 400
+                                           :error  :db/invalid-transaction})))
+          txid-only?    (some-> opts (get "txid-only") str/lower-case (= "true"))
+          auth-id       (:auth auth-map)
```
Comment (zonotope) [inline src/fluree/db/peer/http_api.clj:226]: this is probably outside the scope of this pr, but all the `auth` and `authority` names flying around makes this a bit hard to follow.

**Review by zonotope — APPROVED** [review]

File: src/fluree/db/peer/http_api.clj (hunk 4 above)
Comment (cap10morgan) [reply src/fluree/db/peer/http_api.clj:226]: I agree. Not sure what better names are, and I have `s/db/ledger/g` higher on my priority list before even digging into this one (someday).

**Comment by cap10morgan** [comment]
Talked this over with @bplatz. I'm going to verify that #1 below is true or implement #2 if not:

1. The value we're already storing in the `tx/tx` flake, when hashed w/ SHA-256, matches the digest value in the `tx/signed` flake.
2. Instead of the internal command, we store the original request body in the `tx/tx` flake so that hashing it gives you the digest value in the `tx/signed` flake.

**Comment by cap10morgan** [comment]
@zonotope Sorry, there's a few more commits after your approval to ensure that you can cryptographically recreate the data provenance from what's stored in the flakes. The main change being in  [c8c318a](https://github.com/fluree/ledger/pull/130/commits/c8c318ab8fa75d3d6750047cebfbc2d19f521ff5) to ensure that the value whose SHA-256 hash is in the signing string's `digest` element is stored in the `tx/tx` flake. Mind taking another look?
